chain/chain/src/state_sync.rs

```diff
@@ -0,0 +1,191 @@
+        );
+        return Ok(());
+    }
+
+    let mut done = true;
+    for (shard_index, num_new_chunks) in num_new_chunks.iter_mut().enumerate() {
+        let new_chunk = header.chunk_mask()[shard_index];
+        if new_chunk {
+            *num_new_chunks += 1;
+        }
+        if *num_new_chunks < 2 {
+            done = false;
+        }
+    }
+    if done {
+        // TODO(current_epoch_state_sync): this will not be correct if this block doesn't end up finalized on the main chain.
+        // We should fix it by setting the sync hash when it's finalized, which requires making changes to how we take state snapshots.
+        store_update.set_ser(DBCol::StateSyncHashes, header.epoch_id().as_ref(), header.hash())?;
+        store_update.delete_all(DBCol::StateSyncNewChunks);
+        // TODO: remove old ones
+        return Ok(());
+    }
+
+    store_update.set_ser(DBCol::StateSyncNewChunks, header.hash().as_ref(), &num_new_chunks)?;
+    Ok(())
+}
+
+fn on_new_epoch(store_update: &mut StoreUpdate, header: &BlockHeader) -> Result<(), Error> {
+    let num_new_chunks = vec![0usize; header.chunk_mask().len()];
+    store_update.set_ser(DBCol::StateSyncNewChunks, header.hash().as_ref(), &num_new_chunks)?;
+    Ok(())
+}
+
+fn remove_old_epochs(
+    store: &Store,
+    store_update: &mut StoreUpdate,
+    header: &BlockHeader,
+    prev_header: &BlockHeader,
+) -> Result<(), Error> {
+    for epoch_id in iter_state_sync_hashes_keys(store) {
+        let epoch_id = epoch_id?;
+        if &epoch_id != header.epoch_id() && &epoch_id != prev_header.epoch_id() {
+            store_update.delete(DBCol::StateSyncHashes, epoch_id.as_ref());
+        }
+    }
+    Ok(())
+}
+
+fn update_sync_hashes<T: ChainStoreAccess>(
+    chain_store: &T,
+    store_update: &mut StoreUpdate,
+    header: &BlockHeader,
+) -> Result<(), Error> {
+    let sync_hash = chain_store.get_current_epoch_sync_hash(header.epoch_id())?;
+    if sync_hash.is_some() || header.height() == chain_store.get_genesis_height() {
+        return Ok(());
+    }
+
+    let prev_header = match chain_store.get_block_header(header.prev_hash()) {
+        Ok(h) => h,
+        // During epoch sync, we save headers whose prev headers might not exist, so we just do nothing in this case.
+        // This means that we might not be able to state sync for this epoch, but for now this is not a problem.
+        Err(Error::DBNotFoundErr(_)) => return Ok(()),
+        Err(e) => return Err(e),
+    };
+
+    if prev_header.height() == chain_store.get_genesis_height() {
+        return on_new_epoch(store_update, header);
+    }
+    if prev_header.epoch_id() != header.epoch_id() {
+        on_new_epoch(store_update, header)?;
+        return remove_old_epochs(chain_store.store(), store_update, header, &prev_header);
+    }
+
+    save_epoch_new_chunks(chain_store, store_update, header)?;
+
+    if header.last_final_block() != &CryptoHash::default() {
```

Comment:
nit: Can you move this to a new function?

Reply:
done